keylime/tpm/tpm_abstract.py

```diff
@@ -307,6 +307,8 @@ def check_pcrs(self, agentAttestState, tpm_policy, pcrs, data, virtual, ima_meas
                     return False
                 pcrs_in_quote.add(pcr_num)
 
+        tpm_policy = ast.literal_eval(tpm_policy)
```

Comment:
I think this is actually wrong. Look further above and you will find this here:

```
        try:
            tpm_policy_ = ast.literal_eval(tpm_policy)
        except ValueError:
            tpm_policy_ = {}
```

I think the correct solution is to use `tpm_policy_` below in `if len(list(tpm_policy.keys())) > 0:`. @THS-on ?

Reply:
Using `tpm_policy_` should work but I don't know if it is the actual fix here.
I've looked at the code before the refactoring and there also `tpm_policy` was used which shouldn't work when `tpm_policy` is a string, so this code was never executed before.
https://github.com/keylime/keylime/blob/95ef0415f890b525ab277c37705aba9c22600dff/keylime/tpm/tpm_abstract.py#L292-L298

`len(tpm_policy_.keys()) > 0` will always hold when the policy could be successfully evaluated and contains a mask, so why do we have this check here? I think we can remove it and always warn if we have a PCR in our quote that we haven't used in some way.

The second question what PCR is not in `pcr_in_quote` that causes this now code to executed? Because that might be a regression.

Reply:
Also looking at the failing test we currently assume `tpm_policy`  is always a dict encoded as a string. This seems not always to be the case so we should try to eval it only if it is a string in this part:

https://github.com/keylime/keylime/blob/35d0ae1df571e2719e9338f8ea406d0ed2b4cc6e/keylime/tpm/tpm_abstract.py#L236-L239

Another thing is that the tenant uses `json.dumps(..)` but in the verifier `ast.literal_eval(..)` is used. This might cause problems if the format is not identical.